src/python/pants/engine/legacy/options_parsing.py

```diff
@@ -0,0 +1,48 @@
+# coding=utf-8
+# Copyright 2018 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import (absolute_import, division, generators, nested_scopes, print_function,
+                        unicode_literals, with_statement)
+
+from pants.engine.rules import RootRule, rule
+from pants.engine.selectors import Select
+from pants.init.options_initializer import OptionsInitializer
+from pants.option.options_bootstrapper import OptionsBootstrapper
+from pants.util.objects import datatype
+
+
+class OptionsParseRequest(datatype(['args', 'env'])):
+  """Represents a request for Options computation."""
+
+  @classmethod
+  def create(cls, args, env):
+    assert isinstance(args, (list, tuple))
+    return cls(
+      tuple(args),
+      tuple(env.items() if isinstance(env, dict) else env)
```

Comment:
@cosmicexplorer wrote code for passing dicts across the boundary, at least in one direction, possibly both, but it didn't end up getting committed. Maybe we should revive it and use it?

(Or maybe not, because of the mutability of dicts...)

Reply:
probably not worth it for right now imho. for the purposes of `datatype`, the underlying tuple members need to be hashable. `dict` is not - we'd need something akin to a `frozendict` I think. so in lieu of that, flattening dicts into tuples of items seems just as good to me.

Reply:
The code I wrote for passing dicts across the boundary essentially did exactly this in rust (convert into a tuple of 2-tuples), and just ended up calling a new ffi method that just wrapped the result in `dict()`).

Reply:
Note that this codepath isn't trying to pass things across the boundary: just make them hashable (since dicts aren't hashable).

Reply:
In that case... Time to reach for `frozendict`?

(I was about to reach for it today for the same reason...)

Reply:
docs for `frozendict` say: "Note that a frozendict does not guarantee the immutability of its values, so the utility of hash method is restricted by usage."

so, we'd probably have to roll our own safe `frozendict` impl to be able to reasonably rely on that. not opposed to it if we see more opportunities to utilize something like that, but flattening to tuples works fine for the immediate need here afaict.